### PR TITLE
Cleanup google network utils - WIP

### PR DESCRIPTION
**Pull request by wtgee**
This is a cleanup of some of the network utilities. This is prompted both by the desire to have "real-time" uploads as observations are happening as well as to have a suite of tools that can be used in google compute instances for interacting with the storage.

The main upload from the `observatory` now calls out to the `scripts/process_image_dir.py` script. Changes include the uploading of all files in the directory, not just the fits files, as well as the saving (and thus uploading) of all the jpgs. We could consider making and uploading the timelapse into same directory.

Also allows `solve_field` to handle `fz` without decompression.

Also includes the `clouddb` module for interacting with the cloud sql db. Still need to write a few tests for that but have been using it frequently in PIAA. The PGPASSWORD mechanism will probably update in a few days.

### CODE REVIEW

File: pocs/utils/google/storage.py [hunk 1]
```diff
@@ -88,3 +91,239 @@ def upload(self, local_path, remote_path=None):
...
+                                 bucket='panoptes-survey',
+                                 **kwargs):
+    """Upload an observation directory to google cloud storage.
+
+    This is a convenience function for bulk uploading an observation folder to a
+    bucket. This assumes that observations are placed within `/images/fields`
+    and follow the normal naming convention for observations.
+
+    Note:
+        This requires that the command line utility `gsutil` be installed
```
Comment (jamessynge) [inline pocs/utils/google/storage.py]: It does seem odd that we are taking two approaches to dealing with GCS, esp. authorization. Should we just require that `gcloud login` (or `gcloud ...service_account...`) has been done?

File: pocs/utils/google/storage.py [hunk 2]
```diff
@@ -88,3 +91,239 @@ def upload(self, local_path, remote_path=None):
...
+        field_dir = dir_name.split('/fields/')[-1]
+        remote_path = os.path.normpath(os.path.join(
+            bucket,
+            pan_id,
+            field_dir
+        ))
+
+        # normpath strips the trailing slash so add here so files go in directory
+        destination = 'gs://{}/'.format(remote_path)
+        run_cmd = [gsutil, '-mq', 'cp', '-r', img_path, destination]
```
Comment (jamessynge) [inline pocs/utils/google/storage.py]: I just tried running gsutil locally and it failed with:
```
ERROR: Python 3 and later is not compatible with the Google Cloud SDK. Please use Python version 2.7.x.

If you have a compatible Python interpreter installed, you can use it by setting
the CLOUDSDK_PYTHON environment variable to point to it.
```
How is it working for you?

File: scripts/process_image_dir.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env python
```
Comment (jamessynge) [inline scripts/process_image_dir.py]: The filename is not at all descriptive. How about upload_observation.py?

File: conf_files/pocs.yaml [hunk 4]
```diff
@@ -52,6 +50,18 @@ messaging:
     # Must match ports in peas.yaml.
     cmd_port: 6500
     msg_port: 6510
+
+# Google Network #
+panoptes_network:
+    auth_key: # Update
```
[the remark this answers is not in the document]
Comment (wtgee) [reply conf_files/pocs.yaml]: Yes, I think that is preferred so I will change it so it accepts either. I thought I already had actually but either way I'll make sure it works. I think we should encourage the gcloud usage as you suggest.

File: conf_files/pocs.yaml [hunk 5]
```diff
@@ -52,6 +50,18 @@ messaging:
     # Must match ports in peas.yaml.
     cmd_port: 6500
     msg_port: 6510
+
```
[the remark this answers is not in the document]
Comment (wtgee) [reply conf_files/pocs.yaml]: #564

File: pocs/utils/google/clouddb.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        host (str, optional): Hostname, default localhost.
+        db_user (str, optional): Name of db user, default 'panoptes'.
+        db_name (str, optional): Name of db, default 'postgres'.
+        port (int, optional): DB port.
+
+    Returns:
+        `psycopg2.Connection`: DB connection object.
+    """
+    try:
+        pg_pass = os.environ['PGPASSWORD']
```
[the remark this answers is not in the document]
Comment (wtgee) [reply pocs/utils/google/clouddb.py]: It's set by postgres.

File: pocs/utils/google/storage.py (hunk 1 above)
Comment (wtgee) [reply pocs/utils/google/storage.py]: The `gsutil` is part of `gcloud` so I don't think they are mutually exclusive. 

Agree with the authentication as noted above, so we'll move toward the `gcloud` service key authorization.

File: pocs/utils/google/storage.py (hunk 2 above)
Comment (wtgee) [reply pocs/utils/google/storage.py]: Oh, hm. Perhaps I had this set up earlier under python 2.7. Let me look into my setup. In my head the gsutil had moved up to python 3.

File: scripts/process_image_dir.py (hunk 3 above)
Comment (wtgee) [reply scripts/process_image_dir.py]: I actually changed it from that name. :)  It includes the cleanup in the script, which is also compressing fits files, making timelapse, etc. so it is doing more than just uploading. But agree it's not the best name whereas just uploading is pretty obvious.

File: pocs/utils/google/storage.py (hunk 2 above)
Comment (wtgee) [reply pocs/utils/google/storage.py]: So gcloud and gsutil come from the same place and both require python 2.7. The gsutil is nice for uploading because it transparently handles parallel uploads while the python3 client library does not. We can use the service key to authenticate with either. If we wanted to completely avoid python2.7 we could just use the client libraries, just missing that parallel ability, which is very nice to have.

The gcloud and gsutil can come from the ubuntu package manager, so part of #578 .

File: scripts/process_image_dir.py (hunk 3 above)
Comment (jamessynge) [reply scripts/process_image_dir.py]: Perhaps have the cleaning as an option controlled by flag(s), with uploading as the default behavior.

File: pocs/tests/utils/google/test_google_storage.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+import os
+import re
+import shutil
+
+from pocs.utils.config import load_config
+from pocs.utils.error import GoogleCloudError
+from pocs.utils.google.storage import PanStorage, upload_observation_to_bucket
+
+
+pytestmark = pytest.mark.skipif(
```
[the remark this answers is not in the document]
Comment (wtgee) [reply pocs/tests/utils/google/test_google_storage.py]: Haven't mocked this yet. The idea was to test actual connection but should also be testing process. Will add.

File: scripts/process_image_dir.py (hunk 3 above)
Comment (wtgee) [reply scripts/process_image_dir.py]: The primary purpose of this script is to cleanup the directory. The uploading (and sending of metadata) are the optional arguments to that.

File: pocs/utils/google/clouddb.py [hunk 8]
```diff
@@ -3,42 +3,43 @@
 from warnings import warn
 
 import psycopg2
```
Comment (jamessynge) [inline pocs/utils/google/clouddb.py]: Please change the name of the path & file to something other than pocs/utils/google/clouddb.py. I get that it is being used to access a Google offering of postgres, but it is being accessed using a generic postgres library, so how about just calling it pocs/utils/db/postgres.py?

File: scripts/process_image_dir.py [hunk 9]
```diff
@@ -5,56 +5,80 @@
...
 
-    def _print(msg):
-        if verbose:
-            print(msg)
-
-    _print("Cleaning observation directory")
-    clean_observation_dir(directory, remove_jpgs=False)
+    _print("Cleaning observation directory: {}".format(directory))
+    try:
+        clean_observation_dir(directory, remove_jpgs=False, verbose=verbose, **kwargs)
```
Comment (jamessynge) [inline scripts/process_image_dir.py]: How about adding remove_jpgs to the command line flags, instead of hard-coding it? Could be in another PR.

File: scripts/process_image_dir.py [hunk 10]
```diff
@@ -5,56 +5,80 @@
...
     if upload:
         _print("Uploading to storage bucket")
-        upload_observation_to_bucket(pan_id, directory, include_files='*')
+
+        uploaded_files_fn = os.path.join(directory, 'uploaded_files.txt')
+
+        if os.path.exists(uploaded_files_fn):
+            _print("Files have been uploaded already, skipping")
+        else:
+            img_path = upload_observation_to_bucket(
```
Comment (jamessynge) [inline scripts/process_image_dir.py]: The name `img_path` implies this is the path of a single image. The `include_files='*'` implies that non-image files may be uploaded. Further, `img_path` is then passed to glob, implying that it is not a glob.

File: pocs/utils/google/clouddb.py (hunk 8 above)
Comment (wtgee) [reply pocs/utils/google/clouddb.py]: Funny how it evolved to that point and I didn't even realize it. Will change, thanks.

File: scripts/process_image_dir.py (hunk 10 above)
Comment (wtgee) [reply scripts/process_image_dir.py]: I cleaned this up ab it but it's still not very intelligent. It merely records all files in the directory that were uploaded, assuming the upload was successful.

**Comment by wtgee** [comment]
Growing out of control. I am going to split this PR into three separate ones:

* updates to google storage
* updates for google cloudsql (postgres db connections)
* clean observations script and dirs
